lahja/base.py

```diff
@@ -0,0 +1,169 @@
+        an optional second parameter of :class:`~lahja.misc.BroadcastConfig` to decide
+        where this event should be broadcasted to. By default, events are broadcasted across
+        all connected endpoints with their consuming call sites.
+        """
+        pass
+
+    @abstractmethod
+    async def request(self,
+                      item: BaseRequestResponseEvent[TResponse],
+                      config: Optional[BroadcastConfig] = None) -> TResponse:
+        """
+        Broadcast an instance of :class:`~lahja.misc.BaseRequestResponseEvent` on the event bus and
+        immediately wait on an expected answer of type :class:`~lahja.misc.BaseEvent`. Optionally
+        pass a second parameter of :class:`~lahja.misc.BroadcastConfig` to decide where the request
+        should be broadcasted to. By default, requests are broadcasted across all connected
+        endpoints with their consuming call sites.
+        """
+        pass
+
+    @abstractmethod
+    def subscribe(self,
+                  event_type: Type[TSubscribeEvent],
+                  handler: Callable[[TSubscribeEvent], None]) -> Subscription:
+        """
+        Subscribe to receive updates for any event that matches the specified event type.
+        A handler is passed as a second argument an :class:`~lahja.misc.Subscription` is returned
+        to unsubscribe from the event if needed.
+        """
+        pass
+
+    @abstractmethod
+    async def stream(self,
+                     event_type: Type[TStreamEvent],
+                     num_events: Optional[int] = None) -> AsyncGenerator[TStreamEvent, None]:
+        """
+        Stream all events that match the specified event type. This returns an
+        ``AsyncIterable[BaseEvent]`` which can be consumed through an ``async for`` loop.
+        An optional ``num_events`` parameter can be passed to stop streaming after a maximum amount
+        of events was received.
+        """
+        yield  # type: ignore  # yield statemen convinces mypy this is a generator function
+
+    @abstractmethod
+    async def wait_for(self, event_type: Type[TWaitForEvent]) -> TWaitForEvent:
+        """
+        Wait for a single instance of an event that matches the specified event type.
+        """
+        pass
+
+
+class BaseEndpoint(EndpointAPI):
+    """
+    Base class for endpoint implementations that implements shared/common logic
+    """
+    logger = logging.getLogger('lahja.endpoint.Endpoint')
+
+    #
+    # Common implementations
+    #
+    def __reduce__(self) -> None:  # type: ignore
+        raise NotImplementedError('Endpoints cannot be pickled')
+
+    async def wait_for(self, event_type: Type[TWaitForEvent]) -> TWaitForEvent:
+        """
+        Wait for a single instance of an event that matches the specified event type.
+        """
+        agen = self.stream(event_type, num_events=1)
+        event = await agen.asend(None)
+        await agen.aclose()
+        return event
```

Comment:
Wanted to specifically point out the change to how this method was implemented.

Reply:
I like how that is now implemented :+1: